sdk/include/ds/linked_list.h

```diff
@@ -0,0 +1,642 @@
+		 *
+		 * This is marked `cheri_no_subobject_bounds` because some of our cons
+		 * cell implementations use pointer proxies that rely on the bounds
+		 * provided by `this` (which, in turn, is likely to be
+		 * `cheri_no_subobject_bounds`)
+		 */
+		Cell sentinel __attribute__((__cheri_no_subobject_bounds__));
+
+		__always_inline void init()
+		{
+			sentinel.cell_init();
+		}
+
+		__always_inline bool is_empty()
+		{
+			return linked_list::is_singleton(&sentinel);
+		}
+
+		__always_inline void append(Cell *elem)
+		{
+			linked_list::insert_before(&sentinel, elem);
+		}
+
+		__always_inline void append_new(Cell *elem)
+		{
+			linked_list::insert_new_before(&sentinel, elem);
+		}
+
+		__always_inline void prepend(Cell *elem)
+		{
+			linked_list::insert_before(elem, &sentinel);
+		}
+
+		__always_inline Cell *first()
+		{
+			return sentinel.cell_next();
+		}
+
+		__always_inline Cell *last()
+		{
+			return sentinel.cell_prev();
+		}
+
+		__always_inline bool last_is_eq(Cell *p)
+		{
+			return p->cell_next_is_eq(&sentinel);
+		}
+
+		__always_inline Cell *unsafe_take_first()
+		{
+			Cell *f = sentinel.cell_next();
+			linked_list::unsafe_remove_link(&sentinel, f);
+			return f;
+		}
+
+		__always_inline Cell *take_all()
+		{
+			auto p = linked_list::unsafe_remove(&sentinel);
+			sentinel.cell_init();
+			return p;
+		}
+
+		template<typename F>
+		__always_inline bool search(F f)
+		{
+			return linked_list::search(&sentinel, f);
+		}
+	};
+
+	namespace cell
+	{
+
+		/** Cons cell using two pointers */
+		class Pointer
+		{
+			Pointer *prev, *next;
+
+			public:
+			__always_inline void cell_init()
+			{
+				prev = next = this;
+			}
+
+			__always_inline auto cell_next()
+			{
+				return ds::pointer::proxy::Pointer(next);
+			}
+
+			__always_inline auto cell_prev()
+			{
+				return ds::pointer::proxy::Pointer(prev);
+			}
+		};
+		static_assert(HasPrimOpsInit<Pointer>);
+
+		/**
+		 * Encode a linked list cons cell as a pair of addresses (but present an
+		 * interface in terms of pointers).  CHERI bounds on the returned
+		 * pointers are inherited from the pointer to `this` cons cell.
+		 */
+		class PtrAddr
+		{
+			ptraddr_t prev, next;
+
+			public:
+			/* Primops */
+
+			__always_inline void cell_init()
+			{
+				prev = next = CHERI::Capability{this}.address();
+			}
+
+			__always_inline auto cell_next()
+			{
+				return ds::pointer::proxy::PtrAddr(this, next);
+			}
+
+			__always_inline auto cell_prev()
+			{
+				return ds::pointer::proxy::PtrAddr(this, prev);
+			}
+
+			/*
+			 * Specialized implementations that may be slightly fewer
+			 * instructions than the generic approaches in terms of the primops.
+			 */
+
+			__always_inline bool cell_is_singleton()
+			{
+				return prev == CHERI::Capability{this}.address();
+			}
+
+			__always_inline bool cell_is_doubleton()
+			{
+				return prev == next;
+			}
+
+			__always_inline bool cell_next_is_eq(PtrAddr *p)
```

Comment:
It feels like this would be cleaner if you exposed a `next()` that had an `operator==` (or, since it's the future now, an `operator<=>`) implemented.

Reply:
You're absolutely right, and these methods predate the refactoring of cons cells to use pointer proxies (originally the cells themselves were also the locus of proxying), so I have moved the `operator==` and `operator<=>` down to the pointer proxies and eliminated these.